src/pubsub/ua_pubsub_writer.c

```diff
@@ -97,6 +97,155 @@ UA_DataSetWriterConfig_clear(UA_DataSetWriterConfig *pdsConfig) {
     memset(pdsConfig, 0, sizeof(UA_DataSetWriterConfig));
 }
 
+UA_StatusCode
+UA_Server_enableDataSetWriter(UA_Server *server, const UA_NodeId dataSetWriterIdent) {
+    UA_LOCK(&server->serviceMutex);
+    UA_StatusCode ret = UA_STATUSCODE_BADNOTFOUND;
+    UA_DataSetWriter *dataSetWriter = UA_DataSetWriter_findDSWbyId(server, dataSetWriterIdent);
+
+    if(dataSetWriter)
+    {
+        UA_NodeId wgId = dataSetWriter->linkedWriterGroup;
+        UA_WriterGroup* writerGroup = UA_WriterGroup_findWGbyId(server, wgId);
+        if(writerGroup->state == UA_PUBSUBSTATE_OPERATIONAL || writerGroup->state == UA_PUBSUBSTATE_PREOPERATIONAL) {
+            ret = UA_DataSetWriter_setPubSubState(server, dataSetWriter, UA_PUBSUBSTATE_PREOPERATIONAL,
+                                                    UA_STATUSCODE_GOOD);
+            if(ret != UA_STATUSCODE_GOOD) {
+                UA_LOG_ERROR_WRITER(&server->config.logger, dataSetWriter,
+                                        "Enable DataSetWriter failed - setPubSubState failed");
+            }
+        }
+        else
+        {
+            ret = UA_DataSetWriter_setPubSubState(server, dataSetWriter, UA_PUBSUBSTATE_PAUSED,
+                                                  UA_STATUSCODE_GOOD);
+        }
+    }
+    UA_UNLOCK(&server->serviceMutex);
+    return ret;
+}
+
+static UA_StatusCode setPubSubState_disabled(UA_Server *server,
+                                      UA_DataSetWriter *dataSetWriter) {
+    UA_StatusCode ret = UA_STATUSCODE_GOOD;
+    switch (dataSetWriter->state) {
+        case UA_PUBSUBSTATE_DISABLED:
+            break;
+        case UA_PUBSUBSTATE_PAUSED:
+            dataSetWriter->state = UA_PUBSUBSTATE_DISABLED;
+            //no further action is required
+            break;
+        case UA_PUBSUBSTATE_PREOPERATIONAL:
+        case UA_PUBSUBSTATE_OPERATIONAL:
+            dataSetWriter->state = UA_PUBSUBSTATE_DISABLED;
+            break;
+        case UA_PUBSUBSTATE_ERROR:
+            break;
+        default:
+            UA_LOG_WARNING_WRITER(&server->config.logger, dataSetWriter,
+                                    "Received unknown PubSub state!");
+    }
+
+    return ret;
+}
+
+static UA_StatusCode setPubSubState_paused(UA_Server *server,
+                                    UA_DataSetWriter *dataSetWriter) {
+    UA_StatusCode ret = UA_STATUSCODE_GOOD;
+    switch (dataSetWriter->state){
```

Comment:
This doesn't change that state at all!

Reply:
The paused state does not have any affect defined yet so there is no changes to be done except the state change variable updates